File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
answer = 0
def dfs(n, total, numbers, target): # 현재까지 더한 정수들의 수 n, 정수의 합 total
    global answer
    if n == len(numbers):
        if total == target:
            answer += 1
            return
        return
    dfs(n +1, total + numbers[n], numbers, target)
    dfs(n +1, total - numbers[n], numbers, target)
    

def solution(numbers, target):
    global answer
    
    dfs(0, 0, numbers, target)
    return answer

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    int answer = 0;
    public int solution(int[] numbers, int target) {
        dfs(numbers, target, 0, 0);
        return answer;
    }
    void dfs(int[] numbers, int target, int sum, int count){
        // 모든 숫자를 다 사용하면 리턴
        if(numbers.length == count){ //모든 수 사용
            if(target == sum){ //타겟넘버 도달
                answer++;
                return;
            }    
        }
        else{
            dfs(numbers, target, sum - numbers[count], count+1);
            dfs(numbers, target, sum + numbers[count], count+1);
        }
    }
}